Kuadrant/kuadrantctl
# topology command: persist in local disk

### What 

topology command: persist in local disk in [DOT](https://graphviz.org/doc/info/lang.html) format and SVG format

```
❯ bin/kuadrantctl topology -h
Read kuadrant topology

Usage:
  kuadrantctl topology [flags]

Flags:
  -h, --help               help for topology
  -n, --namespace string   Topology's namespace (default "kuadrant-system")
  -o, --output string      Output file (default "(required)")

Global Flags:
  -v, --verbose   verbose output
```

### Verification steps

* Install kuadrant
```
helm repo add kuadrant https://kuadrant.io/helm-charts/ --force-update
helm install kuadrant kuadrant/kuadrant-operator --namespace kuadrant-system --create-namespace
```
* Deploy kuadrant
```
kubectl -n kuadrant-system apply -f - <<EOF
apiVersion: kuadrant.io/v1beta1
kind: Kuadrant
metadata:
  name: kuadrant
spec: {}
EOF
```
* Read topology
```
make install
```
```
bin/kuadrantctl topology -o ~/tmp/topology -v
```
Output
```
2024-12-19T10:55:52+01:00	DEBUG	reading topology	{"object": {"name":"topology","namespace":"kuadrant-system"}, "error": null}
2024-12-19T10:55:52+01:00	DEBUG	write topology in DOT format to file	{"file": "/home/eguzki/tmp/topology.dot", "error": null}
2024-12-19T10:55:52+01:00	DEBUG	parse DOT graph	{"graph empty": false, "error": null}
2024-12-19T10:55:52+01:00	DEBUG	info graph	{"graph nodenum": 14, "error": null}
2024-12-19T10:55:52+01:00	DEBUG	render graph to SVG	{"buf len": 10743, "error": null}
2024-12-19T10:55:52+01:00	DEBUG	write topology in SVG format to file	{"file": "/home/eguzki/tmp/topology", "error": null}
```

* Check DOT format:
```
cat ~/tmp/topology.dot
```
```dot
digraph  {
	
	n4[label="Authorino\nkuadrant-system/authorino",shape="ellipse"];
	n2[label="ConfigMap\nkuadrant-system/topology",shape="ellipse"];
	n6[fillcolor="#e5e5e5",label="Gateway\ngateway-system/kuadrant-ingressgateway",shape="box",style="filled"];
	n8[fillcolor="#e5e5e5",label="Listener\ngateway-system/kuadrant-ingressgateway#http",shape="box",style="filled"];
	n5[fillcolor="#e5e5e5",label="GatewayClass\nenvoygateway",shape="box",style="filled"];
	n7[fillcolor="#e5e5e5",label="HTTPRoute\ndefault/toystore",shape="box",style="filled"];
	n9[fillcolor="#e5e5e5",label="HTTPRouteRule\ndefault/toystore#rule-1",shape="box",style="filled"];
	n10[fillcolor="#e5e5e5",label="HTTPRouteRule\ndefault/toystore#rule-2",shape="box",style="filled"];
	n1[label="Kuadrant\nkuadrant-system/kuadrant",shape="ellipse"];
	n3[label="Limitador\nkuadrant-system/limitador",shape="ellipse"];
	n6->n8[comment="Gateway -> Listener"];
	n8->n7[comment="Listener -> HTTPRoute"];
	n5->n6[comment="GatewayClass -> Gateway"];
	n7->n9[comment="HTTPRoute -> HTTPRouteRule"];
	n7->n10[comment="HTTPRoute -> HTTPRouteRule"];
	n1->n5[comment="Kuadrant -> GatewayClass"];
	n1->n3[comment="Kuadrant -> Limitador"];
	n1->n4[comment="Kuadrant -> Authorino"];
	
}
```
